src/lightning/data/cache/dataloader.py

```diff
@@ -0,0 +1,196 @@
+
+
+def _equal_items(data_1: Any, data_2: Any) -> bool:
+    data_1_flattened, _ = tree_flatten(data_1)
+    data_2_flattened, _ = tree_flatten(data_2)
+
+    if len(data_1_flattened) != len(data_2_flattened):
+        return False
+
+    return all(_equal_item(d1, d2) for d1, d2 in zip(data_1_flattened, data_2_flattened))
+
+
+def _equal_item(d1, d2) -> bool:
+    if not isinstance(d1, type(d2)):
+        raise False
+    if isinstance(d1, torch.Tensor) and isinstance(d2, torch.Tensor):
+        return torch.equal(d1, d2)
+    return d1 == d2
+
+
+class CacheDataset(Dataset):
+    def __init__(
+        self, dataset: Dataset, cache_dir: Optional[str], chunk_size: Optional[int], compression: Optional[str]
+    ):
+        self._datataset = dataset
+        if cache_dir is None:
+            cache_dir = os.path.join(os.getcwd(), "cache_dir", datetime.now().strftime("%m-%d-%Y-%H-%M"))
+            os.makedirs(cache_dir, exist_ok=True)
+            chunk_size = 2 << 26
+        self.cache = Cache(cache_dir, chunk_size=chunk_size, compression=compression)
+        self.is_deterministic = False
+
+    def __len__(self) -> int:
+        return len(self.cache) if self.cache.filled else len(self._datataset)
+
+    def __getitem__(self, index):
+        data_1 = self.cache[index] if self.cache.filled else self._datataset[index]
+        if not self.cache.filled:
+            if not self.is_deterministic:
+                data2 = self._datataset[index]
+                if not _equal_items(data_1, data2):
+                    raise ValueError(
+                        f"Your dataset items aren't deterministic. Found {data_1} and {data2} for index {index}."
+                        " HINT: Use the `lightning.data.cache.Cache` directly within your dataset."
+                    )
+                self.is_deterministic = True
+            self.cache[index] = data_1
+        return data_1
+
+
+class CacheCollateFn:
+    def __init__(self):
+        self.collate_fn = default_collate
+
+    def __call__(self, items):
+        if all(item is None for item in items):
+            return None
+        return self.collate_fn(items)
+
+
+class _SingleProcessDataLoaderIterPatch(_SingleProcessDataLoaderIter):
+    """This is overriden to inform the cache is done chunking."""
+
+    def _next_data(self):
+        try:
+            return super()._next_data()
+        except StopIteration:
+            for v in self._dataset_fetcher.dataset.__dict__.values():
+                if isinstance(v, Cache):
+                    v.done()
+            raise StopIteration()
+
+
+class _MultiProcessingDataLoaderIterPatch(_MultiProcessingDataLoaderIter):
+    def __init__(self, loader):
+        # Patch PyTorch worker loop
+        from torch.utils.data._utils import worker
+
+        from lightning.data.cache.worker import _worker_loop
+
+        worker._worker_loop = _worker_loop
+        super().__init__(loader)
+
+
+class LightningDataLoader(DataLoader):
```

Comment:
Can you unify this with https://github.com/Lightning-AI/lightning/blob/master/src/lightning/data/datasets/iterable.py#L409 ?

Reply:
In another PR. I don't want to change Lit-GPT just yet.